src/Juvix/Compiler/Core/Language.hs

```diff
@@ -110,38 +142,51 @@ type Type = Node
 
 instance HasAtomicity Node where
   atomicity = \case
-    Var {} -> Atom
-    Ident {} -> Atom
-    Constant {} -> Atom
-    App {} -> Aggregate appFixity
-    BuiltinApp {..} | null _builtinArgs -> Atom
-    BuiltinApp {} -> Aggregate lambdaFixity
-    Constr {..} | null _constrArgs -> Atom
-    Constr {} -> Aggregate lambdaFixity
-    Lambda {} -> Aggregate lambdaFixity
-    Let {} -> Aggregate lambdaFixity
-    Case {} -> Aggregate lambdaFixity
-    Pi {} -> Aggregate lambdaFixity
-    Univ {} -> Atom
-    TypeConstr {} -> Aggregate appFixity
+    NVar {} -> Atom
+    NIdt {} -> Atom
+    NCst {} -> Atom
+    NApp {} -> Aggregate appFixity
+    NBlt BuiltinApp {..} | null _builtinAppArgs -> Atom
+    NBlt BuiltinApp {} -> Aggregate lambdaFixity
+    NCtr Constr {..} | null _constrArgs -> Atom
+    NCtr Constr {} -> Aggregate lambdaFixity
+    NLam Lambda {} -> Aggregate lambdaFixity
+    NLet Let {} -> Aggregate lambdaFixity
+    NCase Case {} -> Aggregate lambdaFixity
+    NPi Pi {} -> Aggregate lambdaFixity
+    NUniv Univ {} -> Atom
+    NTyp TypeConstr {} -> Aggregate appFixity
     Closure {} -> Aggregate lambdaFixity
 
 lambdaFixity :: Fixity
 lambdaFixity = Fixity (PrecNat 0) (Unary AssocPostfix)
 
 instance Eq Node where
   (==) :: Node -> Node -> Bool
-  Var _ idx1 == Var _ idx2 = idx1 == idx2
-  Ident _ sym1 == Ident _ sym2 = sym1 == sym2
-  Constant _ v1 == Constant _ v2 = v1 == v2
-  App _ l1 r1 == App _ l2 r2 = l1 == l2 && r1 == r2
-  BuiltinApp _ op1 args1 == BuiltinApp _ op2 args2 = op1 == op2 && args1 == args2
-  Constr _ tag1 args1 == Constr _ tag2 args2 = tag1 == tag2 && args1 == args2
-  Lambda _ b1 == Lambda _ b2 = b1 == b2
-  Let _ v1 b1 == Let _ v2 b2 = v1 == v2 && b1 == b2
-  Case _ v1 bs1 def1 == Case _ v2 bs2 def2 = v1 == v2 && bs1 == bs2 && def1 == def2
-  Pi _ ty1 b1 == Pi _ ty2 b2 = ty1 == ty2 && b1 == b2
-  Univ _ l1 == Univ _ l2 = l1 == l2
-  TypeConstr _ sym1 args1 == TypeConstr _ sym2 args2 = sym1 == sym2 && args1 == args2
+  NVar (Var _ idx1) == NVar (Var _ idx2) = idx1 == idx2
```

Comment:
I think we have instances for the wrapper types. I suppose that then it is slightly less efficient. Is this instance used in the evaluator?

Reply:
Yes, we use it in the evaluator, but we can do instances for wrapper types and have a separate function`structEq` with the same semantics as (==) which optimises the cases for constants and constructors and falls back to == otherwise.